pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,423 @@
+	// register tracing and return false.
+	numRegisteredStore := r.numRegisteredStore.Load()
+	if numRegisteredStore >= numAllowed {
+		return false
+	}
+
+	if numRegisteredReplica > numRegisteredStore {
+		panic(fmt.Sprintf("numRegisteredReplica(%d) > numRegisteredStore(%d)", numRegisteredReplica, numRegisteredStore))
+	}
+	if numRegisteredReplica > numAllowed {
+		panic(fmt.Sprintf("numRegisteredReplica(%d) > numAllowed(%d)", numRegisteredReplica, numAllowed))
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
+	if registerSucceeded {
+		// Add one unconditionally to the replica count.
+		r.numRegisteredReplica.Add(1)
+	}
+	// Note we can't assert numRegisteredStore <= numAllowed because if the
+	// setting is changed it can be temporarily violated on other replicas.
+	return registerSucceeded
+}
+
+// storeEntry stores the trace and returns the newly created value.
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry, propCtx withContext) *traceValue {
+	tv := traceValue{
+		traced:  te,
+		propCtx: propCtx,
+	}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	tv.mu.seenMsgStorageAppendResp = make(map[raftpb.PeerID]bool)
+	r.m.Store(te.Index, &tv)
+	return &tv
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te, nil)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// MaybeRegister is called on an entry that is about to be proposed. This will
+// begin logging all subsequent updates to this entry. It will return true if
+// the registration succeeded.
+func (r *RaftTracer) MaybeRegister(propCtx withContext, ent raftpb.Entry) bool {
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(propCtx.Context())
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
```

Comment:
Consider placing this check first. It is cheaper than fetching the context/span.

Reply:
Swapped.

Reply:
Done